index.d.ts

```diff
@@ -0,0 +1,54 @@
+// Project: Immutability helper
+// TypeScript Version: 2.2
+
+export = update
+
+// declare function update<T>(data: T, query: update.Query<T>): T
+declare function update<T>(
+  data: ReadonlyArray<T>,
+  query: ArrayOperators<T>,
+): ReadonlyArray<T>
+
+declare function update<T>(data: T, query: Query<T>): object
+declare function update<T>(
+  data: ReadonlySet<T>,
+  query: SetOperators<T>,
+): ReadonlySet<T>
+
+declare function update<K, V>(
+  data: ReadonlyMap<K, V>,
+  query: MapOperators<K, V>,
+): ReadonlyMap<K, V>
+
+type Tree<T> = {[K in keyof T]?: Query<T[K]>}
+type Query<T> =
+  | Tree<T>
+  | ObjectOperators<T>
+  | ArrayOperators<T>
+  | SetOperators<T>
```

Comment:
This would just be `ArrayOperators<any>` and `SetOperators<any>` since `T` here is the type of the object itself, not the type of its elements.